### ISSUE TITLE
List of updates from pytorch master

### ISSUE THREAD

**Issue by fehiepsi** [uber/pyro#1212]
There are several updates from Pytorch master which are relevant to Pyro. So I put them here to have references for future pull requests.

- [x] pytorch has vectorized count Binomial, HalfCauchy, HalfNormal
- [x] pytorch LowRankMultivariateNormal does not have trace_term
- [x] logsumexp is available in Pytorch master
- [x] bo tutorial's code can be more stable after pytorch's lbfgs supporting line search (under reviewing)
- [x] we can remove `matrix_triangular_solve_compat` in Pyro and update SparseGPRegression to deal with trace_term because pytorch's lowrank_mvn does not support it
- [x] `tensor.new()` is depricated, use `.new_empty` instead

Some performance tips:
- [x] use `K.view(-1)[::n+1] += 1` to add identity matrix to a contiguous matrix K
- [x] ~use `x.size(0)` (faster) instead of `x.shape[0]` for a single element retrive from shape~ not important

**Comment by eb8680** [comment]
Also, `torch.jit` has changed significantly, so `pyro.ops.jit` is basically obsolete.

**Comment by fehiepsi** [comment]
@eb8680 Long time ago, you mentioned about a trick to add a term to log_prob from a model. Could you be more explicit about that trick? I am trying to move the `trace_term` [here](https://github.com/uber/pyro/blob/dev/pyro/contrib/gp/models/sgpr.py#L155) outside of `LowRankMultivariateNormal` distribution (to compatible to pytorch master). That `trace_term` is added directly to `log_prob` of lowrank_mvn.

Thanks!

**Comment by eb8680** [comment]
> a trick to add a term to log_prob from a model

You can add an arbitrary factor to a model's trace by observing a Bernoulli distribution:
```python
pyro.sample("factor", dist.Bernoulli(logits=arbitrary_value), obs=1.)
```

**Comment by fehiepsi** [comment]
Thank you @eb8680 ! I still can not write down the math here. I see that `log(sigmoid(logit))` is different to `logit`. I'll write things down again after lunch. Maybe I miss something.

**Comment by eb8680** [comment]
I guess there's a bit of extra plumbing in `dist.Bernoulli`, but there's some manipulation such that `Bernoulli(...).log_prob(1.) == arbitrary_value`, which is what you're trying to achieve.   You could do something similar with `Categorical`.

**Comment by fehiepsi** [comment]
Yeah, got it! Thanks a lot! :)

**Comment by eb8680** [comment]
The JIT is covered by #1063.  Closing now that #1225 is merged.
